tls/s2n_x509_validator.c

```diff
@@ -399,6 +400,65 @@ S2N_RESULT s2n_x509_validator_read_asn1_cert(struct s2n_stuffer *cert_chain_in_s
     return S2N_RESULT_OK;
 }
 
+/**
+* Validates that each certificate in a peer's cert chain contains only signature algorithms in a security policy's
+* certificate_signatures_preference list.
+*/
+S2N_RESULT s2n_validator_check_cert_preferences(struct s2n_connection *conn, X509 *cert)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(cert);
+
+    const struct s2n_security_policy *security_policy = NULL;
+    RESULT_GUARD_POSIX(s2n_connection_get_security_policy(conn, &security_policy));
+
+    /**
+     * We only restrict the signature algorithm on the certificates in the
+     * peer's certificate chain if the certificate_signature_preferences field
+     * is set in the security policy. This is contrary to the RFC, which
+     * specifies that the signatures in the "signature_algorithms" extension
+     * apply to signatures in the certificate chain in certain scenarios, so RFC
+     * compliance would imply validating that the certificate chain signature
+     * algorithm matches one of the algorithms specified in the
+     * "signature_algorithms" extension.
+     *
+     *= https://www.rfc-editor.org/rfc/rfc5246#section-7.4.2
+     *= type=exception
+     *= reason=not implemented due to lack of utility
+     *# If the client provided a "signature_algorithms" extension, then all
+     *# certificates provided by the server MUST be signed by a
+     *# hash/signature algorithm pair that appears in that extension.
+     *
+     *= https://www.rfc-editor.org/rfc/rfc8446#section-4.2.3
+     *= type=exception
+     *= reason=not implemented due to lack of utility
+     *# If no "signature_algorithms_cert" extension is present, then the
+     *# "signature_algorithms" extension also applies to signatures appearing in
+     *# certificates.
+     */
+    if (security_policy->certificate_signature_preferences == NULL) {
+        return S2N_RESULT_OK;
+    }
+
+    struct s2n_cert_info info = { 0 };
+    RESULT_GUARD(s2n_openssl_x509_get_cert_info(cert, &info));
+
+    /* Ensure that the certificate signature does not use SHA-1. While this check
+     * would ideally apply to all connections, we only enforce it when certificate
+     * preferences exist to stay backwards compatible.
+     */
+    if (conn->actual_protocol_version == S2N_TLS13 && !info.self_signed) {
+        RESULT_ENSURE(info.signature_digest_nid != NID_sha1, S2N_ERR_CERT_UNTRUSTED);
+    }
+
+    if (!info.self_signed) {
```

Comment:
Why does the SHA1 check apply to self-signed, unlike the other checks? Isn't that a behavior change?

Reply:
This is not a behavior change, but instead just a refactor. 

- Certificate signature algorithm is in the test certificate signature preferences list but signature is SHA-1 and TLS 1.3 has been negotiated.
- Certificate signature algorithm is in the test certificate signature preferences list and signature is SHA-1 and TLS 1.2 has been negotiated.

Those tests cases confirm that no behavior change occurred. More discussion is in the other comment (which I apparently can't link to until I publish the review)

Reply:
Are you sure? Do we check for SHA1 somewhere other than [here](https://github.com/aws/s2n-tls/blob/6c7cdc8329d846d7fcc8e505cd7708e3f2604d70/tls/s2n_x509_validator.c#L935-L944)? Because it looks like it's checked alongside the preferences in the policy.

Reply:
Oh wait, my bad. You're checking !info.self_signed in BOTH cases. 

It might be clearer if you just return early if self_signed.